force-app/main/default/classes/RollupRecord.cls

```diff
@@ -124,25 +167,253 @@ public with sharing class RollupRecord {
+            childRelationshipName
+        );
+
+        if (childDeliveries == null) {
+            return;
+        }
+
+        for (ServiceDelivery__c delivery : childDeliveries) {
+            existingDeliveries.add(new ServiceDelivery(delivery));
+        }
+    }
+
+    public void addDelivery(ServiceDelivery delivery) {
+        if (deliveries == null) {
+            deliveries = new List<ServiceDelivery>();
+        }
+
+        deliveries.add(delivery);
+    }
+
+    public void addOriginalDelivery(ServiceDelivery delivery) {
+        if (originalDeliveries == null) {
+            originalDeliveries = new List<ServiceDelivery>();
+        }
+
+        originalDeliveries.add(delivery);
+    }
+
+    public void process() {
+        if (originalDeliveries != null && deliveries != null) {
+            processUpdates();
+            return;
+        }
+
+        Boolean isNew = originalDeliveries == null;
+        Boolean reprocessLastServiceDate = false;
+        Boolean reprocessConsecutiveAbsences = false;
+        List<ServiceDelivery> deliveriesToProcess = isNew
+            ? deliveries
+            : originalDeliveries;
+
+        for (ServiceDelivery delivery : deliveriesToProcess) {
+            if (delivery.isPresent()) {
+                reprocessConsecutiveAbsences = true;
+                if (isNew) {
+                    incrementPresent();
+                } else {
+                    decrementPresent();
+                }
+            } else if (delivery.isAbsent()) {
+                reprocessConsecutiveAbsences = true;
+                if (isNew) {
+                    incrementAbsent();
+                } else {
+                    decrementAbsent();
+                }
+            }
+            // Last Service Date is only on Contact and Program Engagement no need to check
+            // when the field is not present
+            if (lastServiceDateField == null) {
+                continue;
+            }
+
+            if (delivery.isDelivered()) {
+                if (
+                    isNew && delivery.getDeliveryDate() > lastServiceDate ||
+                    lastServiceDate == null
+                ) {
+                    lastServiceDate = delivery.getDeliveryDate();
+                } else if (!isNew && delivery.getDeliveryDate() == lastServiceDate) {
+                    lastServiceDate = null;
+                    reprocessLastServiceDate = true;
+                }
+            }
+        }
+
+        if (reprocessLastServiceDate || reprocessConsecutiveAbsences) {
+            processExistingDeliveries(
+                reprocessLastServiceDate,
+                reprocessConsecutiveAbsences
+            );
+        }
+    }
+
+    private void processUpdates() {
+        Map<Id, ServiceDelivery> originalDeliveryById = new Map<Id, ServiceDelivery>();
+        for (ServiceDelivery originalDelivery : originalDeliveries) {
```

Comment:
Can we not use the map method putAll() here instead of looping through the list?

Reply:
No because we are using a wrapper class not the actual sObject here.